### ISSUE TITLE
MockMaker initialization fails post upgrade of spring-boot 2.7.x to 3.1.x

### ISSUE THREAD

**Issue by bpraharaj** [mockito/mockito#3091]
We are updating our project from spring-boot 2.7.14 to 3.1.2 and facing the issue with mockito as "Could not initialize plugin: interface org.mockito.plugins.MockMaker"

Spring-boot: 5.1.2
Java: Azul JDK 17

Spring-boot brings default mockito-core 5.3.1.
Latest version of mockito-inline is 5.2.0, which has a dependency on mockito-core 5.2.0.
Suspecting these library compatibility might be causing this issue, we excluded mockito-core and mockito-junit-jupiter from spring-boot-starter-test. And declared the dependency on mockito-inline 5.2.0 as follows.
![image](https://github.com/mockito/mockito/assets/26647429/8421f974-faf6-4f6f-91c8-5348b0c54e55)

Exception stacktrace:

  java.lang.IllegalStateException: Could not initialize plugin: interface org.mockito.plugins.MockMaker (alternate: null)
  	at org.mockito.internal.configuration.plugins.PluginLoader$1.invoke(PluginLoader.java:84) ~[mockito-core-5.3.1.jar:na]
  	at jdk.proxy2/jdk.proxy2.$Proxy25.getHandler(Unknown Source) ~[na:na]
  	at org.mockito.internal.util.MockUtil.getMockHandlerOrNull(MockUtil.java:158) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.util.MockUtil.isMock(MockUtil.java:147) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.util.DefaultMockingDetails.isMock(DefaultMockingDetails.java:32) ~[mockito-core-5.3.1.jar:na]
  	at org.springframework.boot.test.mock.mockito.MockReset.get(MockReset.java:106) ~[spring-boot-test-3.1.2.jar:3.1.2]
  	at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.resetMocks(ResetMocksTestExecutionListener.java:85) ~[spring-boot-test-3.1.2.jar:3.1.2]
  	at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.resetMocks(ResetMocksTestExecutionListener.java:73) ~[spring-boot-test-3.1.2.jar:3.1.2]
  	at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.beforeTestMethod(ResetMocksTestExecutionListener.java:60) ~[spring-boot-test-3.1.2.jar:3.1.2]
  	at org.springframework.test.context.TestContextManager.beforeTestMethod(TestContextManager.java:288) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:74) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
  	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:252) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) ~[junit-4.13.2.jar:4.13.2]
  	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) ~[junit-4.13.2.jar:4.13.2]
  	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191) ~[spring-test-6.0.11.jar:6.0.11]
  	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.runner.JUnitCore.run(JUnitCore.java:115) ~[junit-4.13.2.jar:4.13.2]
  	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42) ~[junit-vintage-engine-5.9.3.jar:5.9.3]
  	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80) ~[junit-vintage-engine-5.9.3.jar:5.9.3]
  	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72) ~[junit-vintage-engine-5.9.3.jar:5.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86) ~[junit-platform-launcher-1.9.3.jar:1.9.3]
  	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:50) ~[surefire-junit-platform-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:184) ~[surefire-junit-platform-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:148) ~[surefire-junit-platform-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:122) ~[surefire-junit-platform-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385) ~[surefire-booter-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162) ~[surefire-booter-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507) ~[surefire-booter-3.0.0.jar:3.0.0]
  	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495) ~[surefire-booter-3.0.0.jar:3.0.0]
  Caused by: java.lang.IllegalStateException: Failed to load interface org.mockito.plugins.MockMaker implementation declared in java.lang.CompoundEnumeration@71468613
  	at org.mockito.internal.configuration.plugins.PluginInitializer.loadImpl(PluginInitializer.java:56) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.configuration.plugins.PluginLoader.loadPlugin(PluginLoader.java:65) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.configuration.plugins.PluginLoader.loadPlugin(PluginLoader.java:50) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.configuration.plugins.PluginRegistry.<init>(PluginRegistry.java:27) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.configuration.plugins.Plugins.<clinit>(Plugins.java:22) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.internal.MockitoCore.<clinit>(MockitoCore.java:73) ~[mockito-core-5.3.1.jar:na]
  	at org.mockito.Mockito.<clinit>(Mockito.java:1683) ~[mockito-core-5.3.1.jar:na]
  	at org.springframework.boot.test.mock.mockito.MockReset.get(MockReset.java:105) ~[spring-boot-test-3.1.2.jar:3.1.2]
  	... 42 common frames omitted
  Caused by: java.lang.reflect.InvocationTargetException: null
  	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
  	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[na:na]
  	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:na]
  	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[na:na]
  	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[na:na]
  	at org.mockito.internal.configuration.plugins.PluginInitializer.loadImpl(PluginInitializer.java:51) ~[mockito-core-5.3.1.jar:na]
  	... 49 common frames omitted
  Caused by: org.mockito.exceptions.base.MockitoInitializationException: 
  Could not initialize inline Byte Buddy mock maker.
  
  It appears as if your JDK does not supply a working agent attachment mechanism.
  Java               : 17
  JVM vendor name    : Azul Systems, Inc.
  JVM vendor version : 17.0.6+10-LTS
  JVM name           : OpenJDK 64-Bit Server VM
  JVM version        : 17.0.6+10-LTS
  JVM info           : mixed mode, sharing
  OS name            : Linux
  OS version         : 5.4.231-137.341.amzn2.x86_64
  
  		at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.<init>(InlineDelegateByteBuddyMockMaker.java:244)
  		at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.<init>(InlineByteBuddyMockMaker.java:23)
  		... 55 more
  	Caused by: java.lang.IllegalStateException: Error during attachment using: net.bytebuddy.agent.ByteBuddyAgent$AttachmentProvider$Compound@579f3c8e
  		at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:643)
  		at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:616)
  		at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:568)
  		at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:545)
  		at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.<clinit>(InlineDelegateByteBuddyMockMaker.java:115)
  		... 56 more
  	Caused by: java.lang.reflect.InvocationTargetException
  		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  		at net.bytebuddy.agent.Attacher.install(Attacher.java:102)
  		at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:638)
  		... 60 more
  	Caused by: java.io.IOException: Can not attach to current VM
  		at jdk.attach/sun.tools.attach.HotSpotVirtualMachine.<init>(HotSpotVirtualMachine.java:76)
  		at jdk.attach/sun.tools.attach.VirtualMachineImpl.<init>(VirtualMachineImpl.java:57)
  		at jdk.attach/sun.tools.attach.AttachProviderImpl.attachVirtualMachine(AttachProviderImpl.java:58)
  		at jdk.attach/com.sun.tools.attach.VirtualMachine.attach(VirtualMachine.java:207)
  		... 66 more
  	Suppressed: java.lang.IllegalStateException: Could not initialize plugin: interface org.mockito.plugins.MockMaker (alternate: null)
  		at org.mockito.internal.configuration.plugins.PluginLoader$1.invoke(PluginLoader.java:84)
  		at jdk.proxy2/jdk.proxy2.$Proxy25.getHandler(Unknown Source)
  		at org.mockito.internal.util.MockUtil.getMockHandlerOrNull(MockUtil.java:158)
  		at org.mockito.internal.util.MockUtil.isMock(MockUtil.java:147)
  		at org.mockito.internal.util.DefaultMockingDetails.isMock(DefaultMockingDetails.java:32)
  		at org.springframework.boot.test.mock.mockito.MockReset.get(MockReset.java:106)
  		at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.resetMocks(ResetMocksTestExecutionListener.java:85)
  		at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.resetMocks(ResetMocksTestExecutionListener.java:73)
  		at org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener.afterTestMethod(ResetMocksTestExecutionListener.java:67)
  		at org.springframework.test.context.TestContextManager.afterTestMethod(TestContextManager.java:440)
  		at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:94)
  		at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
  		at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
  		... 34 more
  	Caused by: [CIRCULAR REFERENCE: java.lang.IllegalStateException: Failed to load interface org.mockito.plugins.MockMaker implementation declared in java.lang.CompoundEnumeration@71468613]

check that

 - [x] The mockito message in the stacktrace have useful information, but it didn't help
 - [x] The problematic code (if that's possible) is copied here;
       Note that some configuration are impossible to mock via Mockito
 - [x] Provide versions (mockito / jdk / os / any other relevant information)
 - [ ] Provide a [Short, Self Contained, Correct (Compilable), Example](http://sscce.org) of the issue
       (same as any question on stackoverflow.com)
 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)

**Comment by khmarbaise** [comment]
First do you use the spring-boot-parent? Or do you use spring boot via spring-boot-dependencies as a BOM? Apart from that I recommend to define the mockito parts like this (define it in your dependencyManagement):

```xml
      <dependency>
        <groupId>org.mockito</groupId>
        <artifactId>mockito-bom</artifactId>
        <version>5.4.0</version>
        <scope>import</scope>
        <type>pom</type>
      </dependency>
```
and remove a dependency to `mockito-inline` because it's not needed (Also removed since [mockito-5.3.0](https://github.com/mockito/mockito/releases/tag/v5.3.0))... 
That will define all needed artifacts with the correct versions and replace the ones which are coming from spring-boot itself...

Also it is visible that you have mixed different mockito version (via spring-boot mockito 5.3.1 and in your project defined mockito-inline 5.2.0)...

**Comment by bpraharaj** [comment]
@khmarbaise
Sorry for the late response.

I happened to found the root cause of the issue.  Actually, there were no explicit dependency necessary for mockito.  The main culprit was the maven arguments.

The maven argument `-Djdk.attach.allowAttachSelf` was included when we moved our project from Java 8 to Java 17.  Now with spring-boot upgrade, this was causing cyclic dependency and misleading error messages.  We removed the flag `-Djdk.attach.allowAttachSelf` from mvn deploy command, which solved the problem.
